### ISSUE TITLE
Serious lack of quality in documentation

### ISSUE THREAD

**Issue by danielcasler** [fzyzcjy/flutter_rust_bridge#318]
Hello,

I was recently referred to this project by the team at Tokio. It looks very promising, however, the documentation suffers from a serious lack of quality and substance. 

There is no "hello world" example and basic steps to get an example working. The tutorial is beyond brief and contains many broken links. 

We've attempted to run the example projects by cloning the repo and ensuring all the dependencies are installed as per: https://fzyzcjy.github.io/flutter_rust_bridge/quickstart.html#-show-me-the-code

When executing the project using the command: _flutter_rust_bridge_codegen --rust-input frb_example/pure_dart/rust/src/api.rs --dart-output frb_example/pure_dart/lib/bridge_generated.dart --c-output frb_example/pure_dart/ios/Runner/bridge_generated.h_

the process panics with:

`[SEVERE] : Couldn't find dynamic library in default locations.
 [SEVERE] : Please supply one or more path/to/llvm in ffigen's config under the key 'llvm-path'.`

Please provide a simple step-by-step guide to setting up a **hello world** application. Please do not simply write refer to this and refer to that, and then see this over here and download this here and hope it works. This is an incredibly unprofessional way to present your tool and makes it very difficult to take your project seriously as a professional tool.

**Comment by fzyzcjy** [comment]
I agree. This library, at the beginning, was very small and I never thought it will have the popularity today ;) 

And, I usually reply quickly (unless sleeping, surely). So open issues when you have problems as well.

> Please provide a simple step-by-step guide to setting up a hello world application.

Could you plz have a look at: https://fzyzcjy.github.io/flutter_rust_bridge/tutorial_with_flutter.html In that example, just clone the code and run it and it is done.

> [SEVERE] : Couldn't find dynamic library in default locations. [SEVERE] : Please supply one or more path/to/llvm in ffigen's config under the key 'llvm-path'.

That is ffigen's complain, because it cannot understand where your llvm is. llvm is a dependency of ffigen (not this library)

**Comment by Desdaemon** [comment]
Sure, I can do that.

**Comment by fzyzcjy** [comment]
@Desdaemon Sounds great!

**Comment by danielcasler** [comment]
Thank you for the quick reply. 

When cloning the repository and executing **flutter run** from within the frb_example/with_flutter folder, the example works displaying Example 1 and Example 2. 

We are able to generate bindings using **ffigen** without issue when using ffigen on its own. We are able to add a ffigen section to the **pubspec.yaml f**ile:

`ffigen:
  name: NativeLibrary
  description: Bindings to `headers/example.h`.
  output: "generated_bindings.dart"
  llvm-path:
    - "/usr/lib/llvm-13"
  headers:
    entry-points:
      - "headers/example.h"
`
The pubspec.yaml files in the examples do not include ffigen, so we have no idea where to even configure llvm for FRB.

Like I said in my first comment, the documentation is incredibly sparse and not user-friendly. I do understand that this is a small project that has started to take off, so hopefully you guys can clean it up. I do believe it looks promising, and we excited to experimented with FRB.

What you need is a simple step-by-step guide that covers creating the basic project structure, installing dependencies, and performing a hello world. You could also walk the user through adding more complex examples once a user is up and running. 

Your tutorial basically says: _In this tutorial, let us draw a [Mandelbrot set](https://en.wikipedia.org/wiki/Mandelbrot_set), which is plotted in Flutter UI, generated by Rust algorithm, and communicated via this library._ Cool.

It then says clone this repo, run this command. This is not a tutorial. It's cloning something and running it. Not useful.

I'm not trying to be overly critical, and appreciate the work you've put into this, and I hope you can come up with more comprehensive documentation - more inline with the quality one expects from Rust projects.

**Comment by danielcasler** [comment]
When trying to run main.dart:

  RangeError (index): Invalid value: Valid value range is empty: 0

  dart:core-patch/array.dart 121:36                    List.[]
  package:flutter_rust_bridge_example/main.dart 10:27  main.<fn>
  package:flutter_rust_bridge_example/main.dart 9:21   main.<fn>
  package:test_api/src/backend/declarer.dart 215:19    Declarer.test.<fn>.<fn>
  ===== asynchronous gap ===========================
  package:test_api/src/backend/declarer.dart 213:7     Declarer.test.<fn>
  ===== asynchronous gap ===========================
  package:test_api/src/backend/invoker.dart 257:7      Invoker._waitForOutstandingCallbacks.<fn>

**Comment by ramanverma2k** [comment]
I agree, took me a whole day to figure out how to bootstrap a starter flutter project with package. I had to install multiple things at were not even mentioned in the documentation and had to search for the issues separately in order to know what I was missing but It works great when it is configured properly.

**Comment by fzyzcjy** [comment]
@danielcasler

> The pubspec.yaml files in the examples do not include ffigen, so we have no idea where to even configure llvm for FRB.

frb executes ffigen in its flutter_rust_bridge_codegen binary, and you can look at https://github.com/fzyzcjy/flutter_rust_bridge/blob/b9f85cb00dd0694101353e20fd8e3cfcedbe4a0d/frb_codegen/src/commands.rs#L134

> the documentation is incredibly sparse and not user-friendly. I do understand that this is a small project that has started to take off, so hopefully you guys can clean it up. I do believe it looks promising, and we excited to experimented with FRB.

Sure, @Desdaemon and I will improve the documentation soon, and I will come back to this issue when we are finished.

@ramanverma2k

Sorry to hear that. Personally I have installed everything on my computer and cannot uninstall everything to get a brand new env to work with. But I guess the CI files are a good way to refer to, since that installs everything from the beginning. Anyway we will improve doc.

And thank you for appreciating this lib :)

**Comment by fzyzcjy** [comment]
@Desdaemon Hi how is it going on? Maybe you can make a draft PR so I can cooperate when I have some time.

**Comment by ramanverma2k** [comment]
I can also contribute to the documentation. I'll wait for the PR and then go from there.

**Comment by danielcasler** [comment]
I've started putting together an example project with step-by-step instructions. It's very much a WIP.

https://github.com/Neksuhs/rusty_dart/blob/development/STEPS.md

I will likely publish this as a standalone article on Medium - maybe.

**Comment by Desdaemon** [comment]
@danielcasler Looks good! I made a stripped-down version of the pure-dart example and made it into a template, if you want to check it out and see if it works for general usage.

https://github.com/Desdaemon/flutter_rust_bridge_template

**Comment by fzyzcjy** [comment]
@Desdaemon 

> I was planning to create a template repo then write step-by-step instructions around it, so I took some time setting up auto-build for all the platforms. I will create a draft PR this week.

Good idea! GitHub's "template repo"  looks friendly for developers

**Comment by fzyzcjy** [comment]
Cross-linking for who is interested: @Desdaemon has made a wonderful PR - https://github.com/fzyzcjy/flutter_rust_bridge/pull/320

**Comment by danielcasler** [comment]
@Desdaemon Nice. The Flutter example will be a good reference. I didn't just want to give people a project to run, I wanted to give them the basics on how to set up a project from scratch. This way people who are new to FFI get a little more hand-holding at the start. 

My team started with Go, actually, and we've been using Go for years. We recently started with Rust for a few projects and really like it. Between Rust and Go, there is no way I could convince anyone here to use Dart for any sort of serious business logic layer. UI stuff, sure, but asking them to use Dart is such a downgrade = no one would do it and I can't really blame them. lol

We are hoping that the FFI will allow us to build a nice business logic layer similar to AppFlowy. 
https://blog-appflowy.ghost.io/tech-design-flutter-rust/

**Comment by fzyzcjy** [comment]
> We are hoping that the FFI will allow us to build a nice business logic layer similar to AppFlowy.

Yes, frb currently has the capability.

**Comment by danielcasler** [comment]
Is it possible with StreamSink to stream structs to Dart?

**Comment by fzyzcjy** [comment]
Yes. Personally I use streams to send logs (you know, a struct with tag, message, time, ...) from rust to dart.

**Comment by fzyzcjy** [comment]
@danielcasler Sure. My flutter app works fine on both.

**Comment by danielcasler** [comment]
Is there an example of how to use StreamSink with a custom struct?

**Comment by fzyzcjy** [comment]
https://github.com/fzyzcjy/flutter_rust_bridge/blob/3cb8833befcc8702d40ccd5265ed3edc463a2dde/frb_example/pure_dart/rust/src/api.rs?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L171

**Comment by fzyzcjy** [comment]
Just do it as normal https://github.com/fzyzcjy/flutter_rust_bridge/blob/3cb8833befcc8702d40ccd5265ed3edc463a2dde/frb_example/pure_dart/rust/src/api.rs?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L171

**Comment by danielcasler** [comment]
Yeah, I tried that but it wasn't working. The compiler was complaining that the struct didn't implement IntoDart trait.

**Comment by danielcasler** [comment]
I was really surprised by how performant this is. I have it working, streaming multiple structs back to Dart in sub 500 μs

**Comment by fzyzcjy** [comment]
> I am really surprised (and impressed) by how performant this is. I have it working, streaming multiple structs back to Dart in sub 500 μs
I don't know why but I was expecting millisecond latency.

Wow that is great! I guess it is because, frb is really a *very minimal wrapper* (though feature-rich). You can look at the code. It is nothing but a ffi call or some Dart API call.

**Comment by danielcasler** [comment]
It's a great little project once you understand it all. Good job!

I noticed it's using @Desdaemon's project **allo-isolate,** which talks about communicating directly with a Dart isolate using the native port for a send port. Could you tell me what a native port is? The Dart docs don't say much. 

Is it an FFI for talking to an isolate directly, or is it like a raw port that can be used to send bits of data? Sorry for the question, this one is a little outside my experience currently.

**Comment by fzyzcjy** [comment]
> I noticed it's using @Desdaemon's project allo-isolate

Well indeed @shekohex . I contribute to that repo as well, since the original code lacks some parts.

> Could you tell me what a native port is? The Dart docs don't say much.

> Is it an FFI for talking to an isolate directly, or is it like a raw port that can be used to send bits of data? Sorry for the question, this one is a little outside my experience currently.

just like a "handle"?

**Comment by danielcasler** [comment]
The compiler is complaining again about that trait. Can you spot what I've done wrong by chance, please?

`error[E0277]: the trait bound `DartCObject: From<MySize>` is not satisfied
  --> src/api.rs:22:28
   |
22 | pub fn handle_stream(sink: StreamSink<MySize>, arg: String) -> Result<()> {
   |                            ^^^^^^^^^^^^^^^^^^ the trait `From<MySize>` is not implemented for `DartCObject`
   |
   = note: required because of the requirements on the impl of `Into<DartCObject>` for `MySize`
   = note: required because of the requirements on the impl of `IntoDart` for `MySize`
note: required by a bound in `StreamSink`
  --> /home/cazanator/.cargo/registry/src/github.com-1ecc6299db9ec823/flutter_rust_bridge-1.15.1/src/rust2dart.rs:70:26
   |
70 | pub struct StreamSink<T: IntoDart> {
   |                          ^^^^^^^^ required by this bound in `StreamSink``

`#![allow(unused_variables)]

use anyhow::Result;
use flutter_rust_bridge::StreamSink;

#[derive(Debug, Clone)]
pub struct MySize {
    pub width: i32,
    pub height: i32,
}
pub fn handle_stream(sink: StreamSink<MySize>, arg: String) -> Result<()> {
    println!("handle_stream arg={}", arg);

    for i in 0..5 {
        sink.add(MySize {
            width: 20,
            height: 30,
        });
    }
    sink.close();
    Ok(())
}
`

**Comment by danielcasler** [comment]
I noticed you added this to the main README.md file: Fast: It is only a thin (though feature-rich) wrapper, no overhead such as protobuf serialization, thus performant (e.g. <500μs [when](https://github.com/fzyzcjy/flutter_rust_bridge/issues/318#issuecomment-1034451888) streaming multiple structs back).

My test was very crude just to get a sense as to the latency. It certainly was publishable-ready.

What I will do tomorrow is create some actual benchmarks, with code we can publish, so others can run/validate them as well.

**Comment by fzyzcjy** [comment]
> What I will do tomorrow is create some actual benchmarks, with code we can publish, so others can run/validate them as well.

Sounds great! If you want, you can make a PR and put the benchmarks at, e.g., `/frb_benchmark` folder. The README (and possibly the book) can also be updated for that.

**Comment by fzyzcjy** [comment]
> @danielcasler What I will do tomorrow is create some actual benchmarks, with code we can publish, so others can run/validate them as well.

👀 

Curious about the results!

**Comment by danielcasler** [comment]
I've been tied up and haven't gotten to it yet. Will get to it soon!

**Comment by danielcasler** [comment]
I've got the simplified Flutter example and guide pretty much done now as well.

**Comment by danielcasler** [comment]
With regard to my comment above regarding the affinity diagram, if a user wanted to have a long-running Rust layer, a good architecture in a Flutter/Rust scenario might be to have a main isolate launch a secondary isolate with a StreamSink that then launches the Rust layer. The Rust layer could then provide the sink to threads or Tokio tasks that need to communicate to the Flutter layer.

Would it be possible to pass more than one StreamSink to the Rust layer? For example, say a user had two independent Rust modules that needed to send different data types back to the Flutter layer. From the Flutter side, an `initialize` function could be call to start the Rust layer, passing the StreamSinks to the Rust layer.

Something like `pub fn handle_stream(sink: StreamSink<TestStruct>, sink2: StreamSink<String>, arg: String) -> Result<()> {`

**Comment by fzyzcjy** [comment]
> if a user wanted to have a long-running Rust layer, a good architecture in a Flutter/Rust scenario might be to have a main isolate launch a secondary isolate with a StreamSink that then launches the Rust layer. The Rust layer could then provide the sink to threads or Tokio tasks that need to communicate to the Flutter layer.

Hmm maybe no need? Notice that the rust function calls are async. In other words, say you have a rust function `fn compute_heavy_things()` that will run for 10 seconds. Then your dart code can simply call it, and happily do everything else, and the dart code will never be blocked.

**Comment by tmpfs** [comment]
Heads up for anyone trying to set this up on an M1 and get this error using the dart that ships with flutter:

```
Invalid argument(s): Failed to load dynamic library '/opt/homebrew/opt/llvm/lib/libclang.dylib': dlopen(/opt/homebrew/opt/llvm/lib/libclang.dylib, 0x0001): tried: '/opt/homebrew/opt/llvm/lib/libclang.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/usr/lib/libclang.dylib' (no such file), '/opt/homebrew/Cellar/llvm/13.0.0_2/lib/libclang.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/usr/lib/libclang.dylib' (no such file)
#0      _open (dart:ffi-patch/ffi_dynamic_library_patch.dart:12:43)
#1      new DynamicLibrary.open (dart:ffi-patch/ffi_dynamic_library_patch.dart:23:12)
#2      initializeGlobals (package:ffigen/src/header_parser/data.dart:41:33)
#3      initParser (package:ffigen/src/header_parser/parser.dart:48:3)
#4      parse (package:ffigen/src/header_parser/parser.dart:22:3)
#5      main (package:ffigen/src/executables/ffigen.dart:54:19)
#6      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:32)
#7      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)
', /Users/muji/.cargo/registry/src/github.com-1ecc6299db9ec823/flutter_rust_bridge_codegen-1.16.0/src/commands.rs:261:9
```

The solution is to install the arm64 version of dart and put that in PATH instead of the x86_64 version that flutter ships. See: https://github.com/dart-lang/ffigen/issues/260

**Comment by ramanverma2k** [comment]
> Heads up for anyone trying to set this up on an M1 and get this error using the dart that ships with flutter:
> 
> ```
> Invalid argument(s): Failed to load dynamic library '/opt/homebrew/opt/llvm/lib/libclang.dylib': dlopen(/opt/homebrew/opt/llvm/lib/libclang.dylib, 0x0001): tried: '/opt/homebrew/opt/llvm/lib/libclang.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/usr/lib/libclang.dylib' (no such file), '/opt/homebrew/Cellar/llvm/13.0.0_2/lib/libclang.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/usr/lib/libclang.dylib' (no such file)
> #0      _open (dart:ffi-patch/ffi_dynamic_library_patch.dart:12:43)
> #1      new DynamicLibrary.open (dart:ffi-patch/ffi_dynamic_library_patch.dart:23:12)
> #2      initializeGlobals (package:ffigen/src/header_parser/data.dart:41:33)
> #3      initParser (package:ffigen/src/header_parser/parser.dart:48:3)
> #4      parse (package:ffigen/src/header_parser/parser.dart:22:3)
> #5      main (package:ffigen/src/executables/ffigen.dart:54:19)
> #6      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:32)
> #7      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)
> ', /Users/muji/.cargo/registry/src/github.com-1ecc6299db9ec823/flutter_rust_bridge_codegen-1.16.0/src/commands.rs:261:9
> ```
> 
> The solution is to install the arm64 version of dart and put that in PATH instead of the x86_64 version that flutter ships. See: https://github.com/dart-lang/ffigen/issues/260

Yeah, As of now Flutter doesn't ship with arm64 Dart by default. You need to manually download it from [Dart Website](https://dart.dev) and update the path.

**Comment by ramanverma2k** [comment]
> > if a user wanted to have a long-running Rust layer, a good architecture in a Flutter/Rust scenario might be to have a main isolate launch a secondary isolate with a StreamSink that then launches the Rust layer. The Rust layer could then provide the sink to threads or Tokio tasks that need to communicate to the Flutter layer.
> 
> Hmm maybe no need? Notice that the rust function calls are async. In other words, say you have a rust function `fn compute_heavy_things()` that will run for 10 seconds. Then your dart code can simply call it, and happily do everything else, and the dart code will never be blocked.

Can confirm, when you execute the function from inside of the app, it doesn't block the Ui and main thread. You can use your app while the process is running in the background.

**Comment by fzyzcjy** [comment]
@tmpfs Thanks, doc modified https://github.com/Desdaemon/flutter_rust_bridge/commit/00b992fd419c62e2e78e25cca0f82a8649778be9 (will get merged later).

@ramanverma2k 😄

**Comment by tmpfs** [comment]
Some notes on getting this working in Macos flutter that we can add to the docs later.

Once you have the bindings working add the dynamic library to the `Frameworks` group in xcode, be sure to use the `x86_64-apple-darwin` build until flutter has full aarch64 support.

Then ensure it is bundled. Go to `Build Phases > Bundle Framework` and select the `.dylib` file you added to `Frameworks`.

Now you can load the dynamic library for the Macos desktop:

```dart
const base = 'rust_flutter';
final path = Platform.isWindows
    ? '$base.dll'
    : Platform.isMacOS
        ? 'lib$base.dylib'
        : 'lib$base.so';
late final dylib = Platform.isIOS ? DynamicLibrary.process() : DynamicLibrary.open(path);
late final api = RustFlutterImpl(dylib);
```

Replace `rust_flutter` with the name of your library and change `RustFlutterImpl` to the implementation in your generated dart bindings.

A lot of this is quite new to me, so hope it helps somebody else out!

**Comment by tmpfs** [comment]
Anyone got any pointers on setting up on IOS correctly? I am stuck on this error:

```
[VERBOSE-2:ui_dart_state.cc(209)] Unhandled Exception: Invalid argument(s): Failed to lookup symbol 'store_dart_post_cobject': dlsym(RTLD_DEFAULT, store_dart_post_cobject): symbol not found
#0      DynamicLibrary.lookup (dart:ffi-patch/ffi_dynamic_library_patch.dart:34:70)
#1      RustFlutterWire._store_dart_post_cobjectPtr (package:rust_flutter/bridge_generated.dart:116:76)
#2      RustFlutterWire._store_dart_post_cobjectPtr (package:rust_flutter/bridge_generated.dart)
#3      RustFlutterWire._store_dart_post_cobject (package:rust_flutter/bridge_generated.dart:118:41)
#4      RustFlutterWire._store_dart_post_cobject (package:rust_flutter/bridge_generated.dart)
#5      RustFlutterWire.store_dart_post_cobject (package:rust_flutter/bridge_generated.dart:110:12)
#6      FlutterRustBridgeBase._setUpRustToDartComm (package:flutter_rust_bridge/src/basic.dart:35:11)
#7      new FlutterRustBridgeBase (package:flutter_rust_bridge/src/basic.dart:20:5)
#8      new RustFlutterImpl.raw (package:rust_flutter/bridg<…>
```

I have added the C header file to the project and imported it in the primary bridging header, created the universal binary using `cargo lipo` (in fact I had to ignore the aarch64 architecture `cargo lipo --targets x86_64-apple-ios`) setup up the excluded architectures to include `arm64` so it runs on the simulator.

I have also set *Strip Style* to *Debugging Symbols* in case it was getting stripped but it appears to exist when I check with `objdump -t librust_flutter.a`.

I have configured pods with `pod init` and `pod install` and that all seems fine.

What I don't know yet is how to create the `Pods_Runner.framework` - is that critical?

Code is here: https://github.com/tmpfs/rust-flutter

Thanks for any help 🙏

**Comment by fzyzcjy** [comment]
@tmpfs 

@Desdaemon is doing an excellent job on documentations, so you may have a look at the book: https://github.com/Desdaemon/flutter_rust_bridge/blob/doc_fix/book/src/SUMMARY.md

**Comment by fzyzcjy** [comment]
@tmpfs what about "[Integrating with existing projects]" subsection

**Comment by Desdaemon** [comment]
> but sadly the IOS Setup section is empty :(

@tmpfs It's a draft right now, but the only thing I meant to write there is Rust targets specific to iOS development. I'll fill it out soon.

**Comment by tmpfs** [comment]
> > but sadly the IOS Setup section is empty :(
> 
> @tmpfs It's a draft right now, but the only thing I meant to write there is Rust targets specific to iOS development. I'll fill it out soon.
> 
> As for the error, try using the dummy function within `AppDelegate.swift`, that way XCode won't strip the symbols.

Thank you so much @Desdaemon! That was the bit I was missing.

I will likely have to go through this process again once I can get it all working so I hope to help you finish those docs as I learn more about how all the pieces of the puzzle fit together 👍

**Comment by tmpfs** [comment]
So android was pretty straightforward, just needed to install the right NDK as per the docs and `cargo-ndk`. Then I ran:

```
cargo ndk -t armeabi-v7a -t arm64-v8a -o ./android/app/src/main/jniLibs build --release
```

I couldn't use the Android Studio snippet for integrating the build as the project is using Kotlin for Gradle to I will need to rewrite that to Kotlin at some point.

However, integrating with Linux is proving to be very painful due to Flutter using an old version of cmake (`3.10`) and `FetchContent` not being available until a later version (`3.13`).

I have installed much newer versions of `cmake` but getting `flutter` to use them seems impossible, it always uses the version it installs and I can't find a way to make it use an alternative system version of `cmake`.

I tried installing Corrosion globally and using it that way but that also fails.

@fzyzcjy and @Desdaemon any ideas how to get flutter to use my newer version of `cmake` that should allow me to use the `rust.cmake` include file? I tried using `cmake` directly but that fails with this error:

```
Unhandled exception:
RangeError (index): Invalid value: Only valid value is 0: 1
#0      List.[] (dart:core-patch/array.dart:121:36)
#1      main (file:///home/parallels/snap/flutter/common/flutter/packages/flutter_tools/bin/tool_backend.dart:12:37)
#2      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:32)
#3      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)
ninja: build stopped: subcommand failed.
```

FYI I am using Ubuntu 20.04 on ARM64 so maybe I need to try on x86_64.

**Comment by Desdaemon** [comment]
> I tried installing Corrosion globally and using it that way but that also fails.

This is going to cause some issues down the road. I'll try to add more CI setup and investigate this separately.

**Comment by tmpfs** [comment]
I got it working on Linux with ARM64 by ignoring the `rust.cmake` template and adding this to the *bottom* of the main cmake file:

```cmake
set(NATIVE_LIBRARY "${CMAKE_SOURCE_DIR}/../target/aarch64-unknown-linux-gnu/debug/librust_flutter.so")
install(FILES "${NATIVE_LIBRARY}" DESTINATION "${INSTALL_BUNDLE_LIB_DIR}"
  COMPONENT Runtime)
```

With a little conditional for debug and release builds and that should be good enough for Linux 👍

**Comment by danielcasler** [comment]
> > if a user wanted to have a long-running Rust layer, a good architecture in a Flutter/Rust scenario might be to have a main isolate launch a secondary isolate with a StreamSink that then launches the Rust layer. The Rust layer could then provide the sink to threads or Tokio tasks that need to communicate to the Flutter layer.
> 
> Hmm maybe no need? Notice that the rust function calls are async. In other words, say you have a rust function `fn compute_heavy_things()` that will run for 10 seconds. Then your dart code can simply call it, and happily do everything else, and the dart code will never be blocked.

You're referring to making a call to Rust that starts and stops. I was asking if it is possible to make a call to Rust, start an application layer that runs for the life of the app being open for example and have that layer send messages to Flutter when events take place.

**Comment by Desdaemon** [comment]
> if a user wanted to have a long-running Rust layer, a good architecture in a Flutter/Rust scenario might be to have a main isolate launch a secondary isolate with a StreamSink that then launches the Rust layer. The Rust layer could then provide the sink to threads or Tokio tasks that need to communicate to the Flutter layer.

I believe Rust functions are already called in a separate worker thread, so that calls to Rust never blocks and removes the need to spawn isolates. The internal thread pool should be customizable to have more threads IIRC, please correct me if I'm wrong @fzyzcjy.

**Comment by danielcasler** [comment]
@Desdaemon I've not looked at the generated code enough to know. I just assumed that when working with a StreamSink via FRB, launching an isolate to handle the StreamSink and then communicate back to the main isolate would make the most sense.

**Comment by Desdaemon** [comment]
@danielcasler That's fair. The generated code does not make this obvious, because the thread pool implementation of StreamSink lives [here in the Rust runtime module](https://github.com/fzyzcjy/flutter_rust_bridge/blob/3a2ce3d157a79b8c024758ee25e5dd4bdb4c6ab3/frb_rust/src/handler.rs#L163).

**Comment by fzyzcjy** [comment]
@danielcasler I have made the benchmarking into a separate issue #345 to make the issue tracker clearer ;)

**Comment by Desdaemon** [comment]
> I have installed much newer versions of cmake but getting flutter to use them seems impossible, it always uses the version it installs and I can't find a way to make it use an alternative system version of cmake.

@tmpfs The solution was to bump the CMake version in `CMakeLists.txt` to 3.11, which is the first version that supported FetchContent. You can check out [the diff here](https://github.com/fzyzcjy/flutter_rust_bridge/pull/351/files#diff-c7adff697b95bce5f850aba2adf7a49df2ae349bcf8c3d85c51043d0d3eb67d2R33).

**Comment by tmpfs** [comment]
Found a related issue: https://github.com/flutter/flutter/issues/75600

**Comment by tmpfs** [comment]
Ok, I found it on Ubuntu when `flutter` is installed as a `snap`:

```
/snap/flutter/current/usr/bin/cmake --version
```

Is the culprit! So I tried to back it up and symlink but cannot when installed as a snap:

```
mv: cannot move 'cmake' to 'cmake-3.10': Read-only file system
```

So I think I need to remove the flutter snap and try to install it another way on Ubuntu. Then I can hack the `cmake` binary.

Anyone know why `flutter` is shipping such an out of date version of `cmake`? I wonder if something else will break once I manage to use a newer version...
